### PR TITLE
[Storage] Version Level Worm

### PR DESCRIPTION
**Pull request by seanmcc-msft**

### CODE REVIEW

File: sdk/storage/Azure.Storage.sln [hunk 1]
```diff
@@ -131,6 +131,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Azure.Storage.Blo
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "Perf", "Perf", "{FBC7A205-B57A-44E7-A7CA-2E82F79A4D9C}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Azure.Management.Storage", "Microsoft.Azure.Management.Storage\src\Microsoft.Azure.Management.Storage.csproj", "{AD469AF8-0675-4159-A5B7-F486E48B29CE}"
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.sln:134]: TODO remove this once management plane SDK shipped with Version Level Worm, and update https://github.com/Azure/azure-sdk-for-net/blob/master/eng/Packages.Data.props#L182

File: sdk/storage/Azure.Storage.Blobs/src/Models/BlobImmutabilityPolicy.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+{
+    /// <summary>
+    /// Optional parameters for setting the Immutability Policy of a Blob, Blob Snapshot, or Blob Version.
+    /// </summary>
+    public class BlobImmutabilityPolicy
+    {
+        /// <summary>
+        /// The date and time when the Immutability Policy expires.
+        /// </summary>
+        public DateTimeOffset? ExpiriesOn { get; set; }
```
Comment (gapra-msft) [inline sdk/storage/Azure.Storage.Blobs/src/Models/BlobImmutabilityPolicy.cs:16]: typo: ExpiresOn

File: sdk/storage/Azure.Storage.Blobs/tests/VersionLevelWormTests.cs [hunk 3]
```diff
@@ -0,0 +1,899 @@
...
+            : base(async, serviceVersion, null /* RecordedTestMode.Record /* to re-record */)
+        {
+        }
+
+        [Test]
+        [ServiceVersion(Min = BlobClientOptions.ServiceVersion.V2020_06_12)]
+        public async Task SetImmutibilityPolicyAsync()
+        {
+            // Arrange
+            await using DisposingVersionLevelWormContainer vlwContainer = await GetTestVersionLevelWormContainer(TestConfigOAuth);
```
Comment (seanmcc-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/VersionLevelWormTests.cs:36]: Note that we need to use a test account with OAuth enabled

File: sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs [hunk 4]
```diff
@@ -4838,6 +4886,425 @@ private void SetNameFieldsIfNull()
...
+                immutabilityPolicy: immutabilityPolicy,
+                conditions: conditions,
+                async: false,
+                cancellationToken: cancellationToken)
+            .EnsureCompleted();
+
+        /// <summary>
+        /// Sets the Immutability Policy on a Blob, Blob Snapshot, or Blob Version.
+        /// Note that Blob Versioning must be enabled on your storage account, and the blob
+        /// must be in a Container with Version Level Worm enabled to call
```
Comment (gapra-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:4928]: Update the name

File: sdk/storage/Azure.Storage.Blobs/tests/VersionLevelWormTests.cs [hunk 5]
```diff
@@ -0,0 +1,899 @@
...
+            Response<BlobDownloadInfo> downloadResponse = await blob.DownloadAsync();
+
+            // Assert
+            Assert.AreEqual(expectedImmutabilityPolicyExpiry, downloadResponse.Value.Details.ImmutabilityPolicyExpiresOn);
+            Assert.AreEqual(immutabilityPolicy.PolicyMode, downloadResponse.Value.Details.ImmutabilityPolicyMode);
+        }
+
+        [Test]
+        [ServiceVersion(Min = BlobClientOptions.ServiceVersion.V2020_06_12)]
+        public async Task SetImmutibilityPolicyAsync_IfModifiedSince()
```
Comment (gapra-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/VersionLevelWormTests.cs:88]: should this test be called IfUnModifiedSince

Comment (gapra-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/VersionLevelWormTests.cs:88]: we should probably also add a failure test

File: sdk/storage/Azure.Storage.Blobs/src/Models/BlobImmutabilityPolicy.cs (hunk 2 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/src/Models/BlobImmutabilityPolicy.cs:16]: Fixed.

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs [hunk 6]
```diff
@@ -353,6 +355,7 @@ public partial class BlobContainerProperties
         public System.DateTimeOffset? DeletedOn { get { throw null; } }
         public Azure.ETag ETag { get { throw null; } }
         public bool? HasImmutabilityPolicy { get { throw null; } }
+        public bool HasImmutableStorageWithVersioning { get { throw null; } }
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:358]: how is this different from `HasImmutabilityPolicy` line above?

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs [hunk 7]
```diff
@@ -440,6 +445,9 @@ public partial class BlobDownloadDetails
         public string EncryptionKeySha256 { get { throw null; } }
         public string EncryptionScope { get { throw null; } }
         public Azure.ETag ETag { get { throw null; } }
+        public bool HasLegalHold { get { throw null; } }
+        public System.DateTimeOffset? ImmutabilityPolicyExpiresOn { get { throw null; } }
+        public Azure.Storage.Blobs.Models.BlobImmutabilityPolicyMode? ImmutabilityPolicyMode { get { throw null; } }
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:450]: should we use `BlobImmutabilityPolicy` and one property for this?

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs [hunk 8]
```diff
@@ -696,6 +716,9 @@ public partial class BlobItemProperties
         public string EncryptionScope { get { throw null; } }
         public Azure.ETag? ETag { get { throw null; } }
         public System.DateTimeOffset? ExpiresOn { get { throw null; } }
+        public bool HasLegalHold { get { throw null; } }
+        public System.DateTimeOffset? ImmutabilityPolicyExpiresOn { get { throw null; } }
+        public Azure.Storage.Blobs.Models.BlobImmutabilityPolicyMode? ImmutabilityPolicyMode { get { throw null; } }
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:721]: same here.

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs [hunk 9]
```diff
@@ -721,6 +744,11 @@ public partial class BlobLeaseRequestConditions : Azure.RequestConditions
         public BlobLeaseRequestConditions() { }
         public string TagConditions { get { throw null; } set { } }
     }
+    public partial class BlobLegalHoldResult
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:747]: does class name match <Operation>Result pattern? see https://azure.github.io/azure-sdk/general_design.html#model-types

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs [hunk 10]
```diff
@@ -766,6 +794,9 @@ public partial class BlobProperties
         public string EncryptionScope { get { throw null; } }
         public Azure.ETag ETag { get { throw null; } }
         public System.DateTimeOffset ExpiresOn { get { throw null; } }
+        public bool HasLegalHold { get { throw null; } }
+        public System.DateTimeOffset? ImmutabilityPolicyExpiresOn { get { throw null; } }
+        public Azure.Storage.Blobs.Models.BlobImmutabilityPolicyMode? ImmutabilityPolicyMode { get { throw null; } }
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:799]: same here

File: sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs [hunk 11]
```diff
@@ -2502,6 +2510,16 @@ private void SetNameFieldsIfNull()
         /// If the destination blob should be sealed.
         /// Only applicable for Append Blobs.
         /// </param>
+        /// <param name="destinationImmutabilityPolicy">
+        /// Optional <see cref="BlobImmutabilityPolicy"/> to set on the blob.
+        /// Note that is parameter is only applicable to a blob within a container that
+        /// has version level worm enabled.
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:2516]: could we not "use version level worm" anywhere in the codebase?

File: sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs [hunk 12]
```diff
@@ -4854,6 +4902,425 @@ private void SetNameFieldsIfNull()
...
+        /// <returns>
+        /// A <see cref="Response{BlobImmutabilityPolicy}"/>.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
+        public virtual Response<BlobImmutabilityPolicy> SetImmutabilityPolicy(
+            BlobImmutabilityPolicy immutabilityPolicy,
+            BlobRequestConditions conditions = default,
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:4932]: do all conditions apply ?

File: sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs [hunk 13]
```diff
@@ -4854,6 +4902,425 @@ private void SetNameFieldsIfNull()
...
+                try
+                {
+                    scope.Start();
+                    ResponseWithHeaders<BlobSetImmutabilityPolicyHeaders> response;
+
+                    if (async)
+                    {
+                        response = await BlobRestClient.SetImmutabilityPolicyAsync(
+                            timeout: null,
+                            ifUnmodifiedSince: conditions?.IfUnmodifiedSince,
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:5033]: we should settle discussion about what to do with not using all conditional headers first before we add this API.
Depending on the result we should either validate or create SetImmutabilityPolicyOptions.

Having looked at this API, I can see how validation might be tempting.

File: sdk/storage/Azure.Storage.Blobs/tests/VersionLevelWormTests.cs (hunk 3 above)
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/tests/VersionLevelWormTests.cs:36]: can we not use `vlw` or `VersionLevelWorm` in tests either? We should use official naming everywhere. In few years nobody will remember what `vlw` is.

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs (hunk 6 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:367]: GET https://seanoauthstage.blob.core.windows.net/test-container-9d4eb5b6-8be5-4d1e-ba83-ef2c60f8199e?restype=container HTTP/1.1
x-ms-version: 2020-08-04
Accept: application/xml
x-ms-client-request-id: 2fc22202-060c-4a0b-be5b-9d9e5410f387
x-ms-return-client-request-id: true
User-Agent: azsdk-net-Storage.Blobs/12.9.0-alpha.20210604.1 (.NET Framework 4.8.4300.0; Microsoft Windows 10.0.19043 )
x-ms-date: Fri, 04 Jun 2021 18:06:23 GMT
Authorization: [redacted]
traceparent: 00-5d424cc99718c34f9a124b2cf9e42577-3f00ee44eb443749-00
Host: seanoauthstage.blob.core.windows.net

HTTP/1.1 200 OK
Transfer-Encoding: chunked
Last-Modified: Fri, 04 Jun 2021 18:06:22 GMT
ETag: "0x8D92783760EC0E8"
Server: Windows-Azure-Blob/1.0 Microsoft-HTTPAPI/2.0
x-ms-request-id: 4e6de559-c01e-0030-466c-59b723000000
x-ms-client-request-id: 2fc22202-060c-4a0b-be5b-9d9e5410f387
x-ms-version: 2020-08-04
x-ms-lease-status: unlocked
x-ms-lease-state: available
**x-ms-has-immutability-policy: false**
x-ms-has-legal-hold: false
**x-ms-immutable-storage-with-versioning-enabled: true**
x-ms-blob-public-access: container
x-ms-default-encryption-scope: $account-encryption-key
x-ms-deny-encryption-scope-override: false
Date: Fri, 04 Jun 2021 18:06:22 GMT

0

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs (hunk 7 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:460]: We did not group any other properties in `BlobProperties` in the past.  For example, the copy-related and access tier-related properties are all flat.

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs (hunk 8 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:732]: Same as above.

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs (hunk 10 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:810]: Same as above.

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs (hunk 9 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:758]: Let's let One SDK decide in design review.

File: sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs (hunk 7 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/api/Azure.Storage.Blobs.netstandard2.0.cs:460]: Also, we can let One SDK decide in design review on Monday.

File: sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs (hunk 12 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:4941]: Only `IfUnmodifiedSince` applies.

File: sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs (hunk 13 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:5042]: I agree we should schedule that discussion, but maybe not in this PR.

File: sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs (hunk 11 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:2517]: Fixed.

File: sdk/storage/Azure.Storage.Blobs/tests/VersionLevelWormTests.cs (hunk 3 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/tests/VersionLevelWormTests.cs]: Fixed.

File: sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs [hunk 14]
```diff
@@ -4833,6 +4881,432 @@ private void SetNameFieldsIfNull()
...
+        /// Optional <see cref="CancellationToken"/> to propagate
+        /// notifications that the operation should be cancelled.
+        /// </param>
+        /// <returns>
+        /// A <see cref="Response{BlobImmutabilityPolicy}"/>.
+        /// </returns>
+        /// <remarks>
+        /// A <see cref="RequestFailedException"/> will be thrown if
+        /// a failure occurs.
+        /// </remarks>
```
Comment (kasobol-msft) [inline sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:4908]: Might be worth mentioning which conditions work in <remarks>

**Review by kasobol-msft — APPROVED** [review]

File: sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs (hunk 14 above)
Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:4908]: Fixed.

Comment (kasobol-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:4908]: Well, we should do that on public APIs that call into this internal as well.

Comment (seanmcc-msft) [reply sdk/storage/Azure.Storage.Blobs/src/BlobBaseClient.cs:4908]: My bad, fixed!  I intended to add the comment on the public APIs but forgot.

**Pull request merged**